python/lib/align.py

```diff
@@ -0,0 +1,152 @@
+# Returns the alignment of two sequences.
+
+import numpy as np
+from numba import njit
+
+from lib.sequence import *
+
+# Return the edit distance between two sequences.
+# See `align` below which also returns the aligned strings.
+# This is copied from the c++ implementation in utils.hpp.
+@njit
+def edit_distance(s1: Sequence, s2: Sequence, color=True):
+    # This should only be used for relatively small sequences.
+    l1 = s1.len()
+    l2 = s2.len()
+    assert l1 * l2 <= 1000000
+
+    if l1 == 0:
+        return l2
+    if l2 == 0:
+        return l1
+
+    costs = np.zeros((l1 + 1, l2 + 1), dtype=np.int32)
+    for i in range(l2 + 1):
+        costs[0][i] = i
+
+    for i, c1 in enumerate(s1.seq):
+        costs[i + 1][0] = i + 1
+
+        for j, c2 in enumerate(s2.seq):
+            if c1 == c2:
+                costs[i + 1][j + 1] = costs[i][j]
+            else:
+                costs[i + 1][j + 1] = 1 + min(costs[i][j], costs[i + 1][j], costs[i][j + 1])
+
+    return costs[l1][l2]
+
+
+# Given two sequences, find the pairwise matching exons.
```

Comment:
A bit of abuse of language :) No guarantee that the matching sequences are exons. For all we know, we may match an exons with an intron.

Isn't this code just walking backwards in the Levensthein matrix and aligns the sequences?